File: solutions/problem_1985_find_the_kth_largest_integer_in_the_array/solution.py
"""
Problem 1985: Find the Kth Largest Integer in the Array
=====================================================
Difficulty: Medium
Tags: Array, String, Divide and Conquer, Sorting, Heap (Priority Queue), Quickselect
Pattern: Quickselect

Time Complexity:  O(n) average
Space Complexity: O(1)
"""

class Solution:
    def kthLargestNumber(self, nums: List[str], k: int) -> str:
        # Quickselect - O(n) average time
        import random
        def quickselect(arr, k):
            if len(arr) == 1:
                return arr[0]
            pivot = random.choice(arr)
            lows = [x for x in arr if x < pivot]
            highs = [x for x in arr if x > pivot]
            pivots = [x for x in arr if x == pivot]
            if k < len(lows):
                return quickselect(lows, k)
            elif k < len(lows) + len(pivots):
                return pivot
            else:
                return quickselect(highs, k - len(lows) - len(pivots))
        
        k = k if isinstance(k, int) else 1
        return quickselect(nums, len(nums) - k)

File: solutions/problem_1985_find_the_kth_largest_integer_in_the_array/solution.cpp
/*
 * Problem 1985: Find the Kth Largest Integer in the Array
 * =====================================================
 * Difficulty: Medium
 * Tags: Array, String, Divide and Conquer, Sorting, Heap (Priority Queue), Quickselect
 * Pattern: Quickselect
 *
 * Time Complexity:  O(n) average
 * Space Complexity: O(1)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    string kthLargestNumber(vector<string>& nums, int k) {
        // Quickselect - O(n) average time
        int k = k;
        nth_element(nums.begin(), nums.begin() + nums.size() - k, nums.end());
        return nums[nums.size() - k];
    }
};
